### PR TITLE
feat(OptimizelyConfig): add new fields to OptimizelyConfig

### PR DESCRIPTION
**Pull request by jaeopt**
The following new public properties are added to OptimizelyConfig:
- sdkKey
- environmentKey
- attributes
- audiences
- events
- experimentRules and deliveryRules to OptimizelyFeature
- audiences to OptimizelyExperiment

### CODE REVIEW

File: Sources/Data Model/Experiment.swift [hunk 1]
```diff
@@ -63,4 +75,59 @@ extension Experiment {
...
+    ///
+    /// example:
+    /// - string: "(AUDIENCE(1) OR AUDIENCE(2)) AND AUDIENCE(3)"
+    /// - replaced: "(\"us\" OR \"female\") AND \"adult\""
+    ///
+    /// - Parameter string: before replacement
+    /// - Returns: string after replacement
+    func replaceAudienceIdsWithNames(string: String, audiencesMap: [String: String]) -> String {
+        let beginWord = "AUDIENCE("
+        let endWord = ")"
```
Comment (yasirfolio3) [inline Sources/Data Model/Experiment.swift:96]: just curious, is it possible for user to set such characters as values for condition matching?

File: Sources/Optimizely/OptimizelyConfig.swift [hunk 2]
```diff
@@ -61,21 +86,55 @@ struct OptimizelyConfigImp: OptimizelyConfig {
...
-            return updateVariableData(experiment: $0, features: project.featureFlags)
+        let updatedExperiments = projectConfig.allExperiments.map { experiment -> Experiment in
+            let feature = project.featureFlags.filter({ $0.experimentIds.contains(experiment.id) }).first
+            return updateExperiment(experiment: experiment, feature: feature, audiencesMap: audiencesMap)
+        }
+        
+        let updatedRollouts = projectConfig.project.rollouts.map { rollout -> Rollout in
+            let feature = project.featureFlags.filter({ $0.rolloutId == rollout.id }).first
+            
+            var rollout = rollout
```
Comment (yasirfolio3) [inline Sources/Optimizely/OptimizelyConfig.swift:129]: can we name this better? this is a bit confusing?

**Review by yasirfolio3 — COMMENTED** [review]
Minor suggestions

File: Sources/Optimizely/OptimizelyConfig.swift [hunk 3]
```diff
@@ -61,21 +86,55 @@ struct OptimizelyConfigImp: OptimizelyConfig {
...
+                audiences.append(oldAudience)
+            }
+        }
+        self.audiences = audiences
+        
+        // update experiment data:
+        // - copy feature's variable data to variables in all variations
+        // - serialize experiment audiences to a string
+        
+        // prepare an audience [id: name] mapping for audicens serialization
```
Comment (yasirfolio3) [inline Sources/Optimizely/OptimizelyConfig.swift:118]:
```suggestion
        // prepare an audience [id: name] mapping for audience serialization
```

File: Tests/TestData/optmizelyConfig/optimizely_config_datafile.json [hunk 4]
```diff
@@ -243,10 +318,30 @@
             "id": "12115595439"
         }
     ],
-    "attributes": [],
+    "typedAudiences": [
+      {"id": "20348452263", "conditions": ["and", ["or", ["or", {"value": 18, "type": "custom_attribute", "name": "age", "match": "gt"}]]], "name": "adult"},
+      {"id": "20348352569", "conditions": ["and", ["or", ["or", {"value": 18, "type": "custom_attribute", "name": "age", "match": "lt"}]]], "name": "kid"}
+    ],
+    "audiences": [
```
Comment (yasirfolio3) [inline Tests/TestData/optmizelyConfig/optimizely_config_datafile.json:325]: Can we have 1 or 2 unique audiences in both `typedAudiences` and `audiences` to verify if they are being merged correctly?

**Review by yasirfolio3 — APPROVED** [review]
LGTM, just some minor nits.

File: Sources/Data Model/Experiment.swift [hunk 5]
```diff
@@ -63,4 +75,63 @@ extension Experiment {
...
+    
+    /// Replace audience ids with audience names
+    ///
+    /// example:
+    /// - string: "(AUDIENCE(1) OR AUDIENCE(2)) AND AUDIENCE(3)"
+    /// - replaced: "(\"us\" OR \"female\") AND \"adult\""
+    ///
+    /// - Parameter string: before replacement
+    /// - Returns: string after replacement
+    func replaceAudienceIdsWithNames(string: String, audiencesMap: [String: String]) -> String {
```
Comment (msohailhussain) [inline Sources/Data Model/Experiment.swift:94]: can we have unit test for this method.

**Review by msohailhussain — APPROVED** [review]
lgtm.

File: Sources/Data Model/Experiment.swift (hunk 5 above)
Comment (jaeopt) [reply Sources/Data Model/Experiment.swift:94]: We have a unit test case covering this audience mapping - https://github.com/optimizely/swift-sdk/blob/d91c49c41a9b921604527a4beb6aa9c1c08f105d/Tests/OptimizelyTests-DataModel/ExperimentTests.swift#L152

File: Tests/TestData/optmizelyConfig/optimizely_config_datafile.json [hunk 6]
```diff
@@ -1,7 +1,65 @@
 {
```
Comment (keppel2) [inline Tests/TestData/optmizelyConfig/optimizely_config_datafile.json:1]: optmizelyConfig is spelled wrong.

**Pull request merged**